src/httpclient/auth.rs

```diff
@@ -0,0 +1,176 @@
+use std::fmt::Display;
+
+use crate::data::renku_url::RenkuUrl;
+use ::reqwest as rqw;
+use iso8601_timestamp::{Duration, Timestamp};
+use openidconnect::core::*;
+use openidconnect::reqwest::async_http_client;
+use openidconnect::*;
+use serde::{Deserialize, Serialize};
+use snafu::{ResultExt, Snafu};
+
+// Obtain the device_authorization_url from the OIDC metadata provider.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+struct DeviceEndpointProviderMetadata {
+    device_authorization_endpoint: DeviceAuthorizationUrl,
+}
+impl AdditionalProviderMetadata for DeviceEndpointProviderMetadata {}
+type DeviceProviderMetadata = ProviderMetadata<
+    DeviceEndpointProviderMetadata,
+    CoreAuthDisplay,
+    CoreClientAuthMethod,
+    CoreClaimName,
+    CoreClaimType,
+    CoreGrantType,
+    CoreJweContentEncryptionAlgorithm,
+    CoreJweKeyManagementAlgorithm,
+    CoreJwsSigningAlgorithm,
+    CoreJsonWebKeyType,
+    CoreJsonWebKeyUse,
+    CoreJsonWebKey,
+    CoreResponseMode,
+    CoreResponseType,
+    CoreSubjectIdentifierType,
+>;
+
+pub type TokenResponse = StandardTokenResponse<
+    IdTokenFields<
+        EmptyAdditionalClaims,
+        EmptyExtraTokenFields,
+        CoreGenderClaim,
+        CoreJweContentEncryptionAlgorithm,
+        CoreJwsSigningAlgorithm,
+        CoreJsonWebKeyType,
+    >,
+    CoreTokenType,
+>;
+
+pub fn access_token(r: &TokenResponse) -> String {
+    r.access_token().secret().to_string()
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Response {
+    pub created_at: Timestamp,
+    pub response: TokenResponse,
+}
+impl Response {
+    pub fn expires_at(&self) -> Option<Timestamp> {
+        match &self.response.expires_in() {
+            Some(d) => self
+                .created_at
+                .checked_add(Duration::nanoseconds(d.as_nanos() as i64)),
+            None => None,
+        }
+    }
+}
+
+impl Display for Response {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self.expires_at() {
+            Some(end) => write!(f, "Login successful. Valid until {}.", end.format()),
+            None => write!(f, "Login successful."),
+        }
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct UserCode {
+    pub authorization_url: rqw::Url,
+    pub user_code: String,
+    metadata: DeviceProviderMetadata,
+    device_auth_resp: CoreDeviceAuthorizationResponse,
+}
+impl Display for UserCode {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "Please visit this url and authorize this application:\n{}\nUser-Code: {}",
```

Comment:
suggestion: We should be more explicit in the message that the user should enter this code in the URL.

Reply:
In most cases, the code is already passed via the request parameter and the user is not bothered. But in some cases it might not and the user would then need to enter it. I'll adapt the wording to make it clearer.